Application / Engine boot refactor

dgeb opened the pull request:
Common boot concerns for engine instances and application instances have been moved to the `EngineInstance` class. These concerns include running engine initializers and engine instance initializers. 

Engine initializers are run once before the first engine instance is built, following the [approach taken in ember-engines](https://github.com/dgeb/ember-engines/commit/a7f3a891a4f243e97cc5c1b2d94b2d36b45505bb).

Setup of an engine instance's registry based upon its parent will be handled in a separate PR.

dgeb committed:
Move common boot concerns to EngineInstance.

dgeb committed:
Ensure that Engine initializers are run.

Engine initializers should be run once, and only once, before the first
engine instance is built.

dgeb committed:
Introduce tests for EngineInstance initializers.

chadhietala commented on packages/ember-application/tests/system/engine_instance_initializers_test.js:
```diff
@@ -0,0 +1,366 @@
+});
+
+QUnit.test('initializers are per-engine', function() {
+  expect(0);
```
There might be a better assertion for this.

dgeb replied:
I based that on the analogous [application instance initializer test](https://github.com/emberjs/ember.js/blob/master/packages/ember-application/tests/system/instance_initializers_test.js#L320-L333).

I think using `expect(0)` to verify no error is raised is used pretty often, but I'm open to another alternative. I suppose we should change both tests if we change this one.

dgeb replied:
I'll add some verification that initializers raise if duplicated, which would make both tests more rigorous.

chadhietala replied:
Sounds good

dgeb committed:
Improve testing for uniquely named initializers.

Tests now verify that initializers can be named the same on
_different_ apps/engines, but must be uniquely named on _each_
app/engine.

dgeb commented:
@chadhietala Thanks for the review. Just updated with some more rigorous testing for uniqueness of initializer names.

chadhietala commented:
@dgeb LGTM. Great stuff! ✨

rwjblue commented:
Thanks @dgeb!
